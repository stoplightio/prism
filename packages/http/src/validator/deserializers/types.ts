import { HttpParamStyles } from '@stoplight/types';
import { IHttpNameValue, IHttpNameValues, JSONSchema } from '../../types';

export type deserializeFn<Parameters> = (
  name: string,
  parameters: Parameters,
  schema?: JSONSchema,
  explode?: boolean
) => unknown;

export interface IHttpParamDeserializerRegistry<Parameters, S = HttpParamStyles> {
  get(style: S): IHttpParamStyleDeserializer<Parameters> | undefined;
}

export interface IHttpParamStyleDeserializer<Parameters, S = HttpParamStyles> {
  supports: (style: S) => boolean;
  deserialize: deserializeFn<Parameters>;
}

export type IHttpHeaderParamStyleDeserializer = IHttpParamStyleDeserializer<IHttpNameValue>;
export type IHttpQueryParamStyleDeserializer = IHttpParamStyleDeserializer<IHttpNameValues>;
