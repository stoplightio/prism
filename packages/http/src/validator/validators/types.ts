import { IPrismDiagnostic } from '@stoplight/prism-core/src/types';
import { JSONSchema } from 'http/src/types';

export interface IHttpValidator<Target, Specs> {
  validate(target: Target, specs: Specs[], mediaType?: string): IPrismDiagnostic[];
}

export interface ISchemaValidator<S = JSONSchema> {
  validate(content: any, schema: S): IPrismDiagnostic[];
  supports(mediaType: string): boolean;
}

export interface IValidatorRegistry {
  get(mediaType: string): ((content: any, schema: JSONSchema) => IPrismDiagnostic[]) | undefined;
}
