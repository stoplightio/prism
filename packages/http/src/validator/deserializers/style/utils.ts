import { Dictionary } from '@stoplight/types';

export function createObjectFromKeyValList(items: string[]) {
  return items.reduce((obj: Dictionary<unknown>, item, i) => {
    if (i % 2 === 0) {
      obj[item] = undefined;
    } else {
      obj[items[i - 1]] = item;
    }

    return obj;
  }, {});
}
