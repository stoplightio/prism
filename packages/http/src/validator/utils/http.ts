import { IHttpContent } from '@stoplight/types';

export function resolveContent(content: { [mediaType: string]: IHttpContent }, mediaType?: string) {
  return content[mediaType || '*'] || content['*'];
}
