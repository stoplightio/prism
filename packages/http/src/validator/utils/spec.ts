import { IHttpOperationResponse } from '@stoplight/types';
import { head } from 'fp-ts/lib/Array';
import { Option } from 'fp-ts/lib/Option';
import { NonEmptyArray } from 'fp-ts/lib/NonEmptyArray';

export function findOperationResponse(
  responseSpecs: NonEmptyArray<IHttpOperationResponse>,
  statusCode: number
): Option<IHttpOperationResponse> {
  const sortedSpecs = responseSpecs
    .filter(
      spec => new RegExp(`^${spec.code.replace(/X/g, '\\d')}$`).test(String(statusCode)) || spec.code === 'default'
    )
    .sort((s1, s2) => {
      if (s1.code === 'default') {
        return 1;
      }

      if (s2.code === 'default') {
        return -1;
      }

      return s1.code.split('X').length - s2.code.split('X').length;
    });

  return head(sortedSpecs);
}
