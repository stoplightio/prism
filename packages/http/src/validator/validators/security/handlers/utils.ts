import { DiagnosticSeverity } from '@stoplight/types';
import { Either, left, right } from 'fp-ts/lib/Either';
import { IPrismDiagnostic } from '@stoplight/prism-core';
import { IHttpRequest } from '../../../../types';

export type ValidateSecurityFn = (
  input: Pick<IHttpRequest, 'headers' | 'url'>,
  name: string
) => Either<IPrismDiagnostic, unknown>;

export function genRespForScheme(
  isSchemeProper: boolean,
  isCredsGiven: boolean,
  msg: string
): Either<IPrismDiagnostic, unknown> {
  if (isSchemeProper) {
    return when(isCredsGiven, undefined);
  }

  return left(genUnauthorisedErr(msg));
}

export const genUnauthorisedErr = (msg?: string): IPrismDiagnostic => ({
  severity: DiagnosticSeverity.Error,
  message: 'Invalid security scheme used',
  code: 401,
  tags: msg ? [msg] : [],
});

export function isScheme(shouldBeScheme: string, authScheme: string) {
  return authScheme.toLowerCase() === shouldBeScheme;
}

export function when(condition: boolean, errorMessage: string | undefined): Either<IPrismDiagnostic, unknown> {
  return condition ? right(true) : left(genUnauthorisedErr(errorMessage));
}
