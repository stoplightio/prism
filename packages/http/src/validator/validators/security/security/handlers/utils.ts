import { DiagnosticSeverity, IHttpOperation } from '@stoplight/types';
import { Either, left, right } from 'fp-ts/lib/Either';
import { IPrismDiagnostic } from '@stoplight/prism-core';

export function genRespForScheme(
  isSchemeProper: boolean,
  isCredsGiven: boolean,
  resource: IHttpOperation,
  msg: string
): Either<IPrismDiagnostic, IHttpOperation> {
  if (isSchemeProper) {
    return when(isCredsGiven, undefined, resource);
  }

  return left(genUnauthorisedErr(msg));
}

export const genUnauthorisedErr = (msg?: string): IPrismDiagnostic => ({
  severity: DiagnosticSeverity.Error,
  message: 'Invalid security scheme used',
  code: 401,
  tags: msg ? [msg] : [],
});

export function isScheme(shouldBeScheme: string, authScheme: string) {
  return authScheme.toLowerCase() === shouldBeScheme;
}

export function when(
  condition: boolean,
  errorMessage: string | undefined,
  resource: IHttpOperation
): Either<IPrismDiagnostic, IHttpOperation> {
  return condition ? right(resource) : left(genUnauthorisedErr(errorMessage));
}
