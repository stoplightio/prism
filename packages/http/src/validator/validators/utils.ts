import { IPrismDiagnostic } from '@stoplight/prism-core';
import { DiagnosticSeverity, Segment } from '@stoplight/types';
import { getSemigroup } from 'fp-ts/lib/NonEmptyArray';
import { getValidation } from 'fp-ts/lib/Either';
import { option, tryCatch, Option } from 'fp-ts/lib/Option';
import { sequenceT } from 'fp-ts/lib/Apply';
import * as Ajv from 'ajv';
import { JSONSchema } from '../../';
import * as AjvOAI from 'ajv-oai';

const ajv = new AjvOAI({ allErrors: true, messages: true, schemaId: 'auto' });

export const convertAjvErrors = (errors: Ajv.ErrorObject[] | undefined | null, severity: DiagnosticSeverity) => {
  if (!errors) {
    return [];
  }

  return errors.map<IPrismDiagnostic & { path: Segment[] }>(error => {
    const allowedParameters = 'allowedValues' in error.params ? `: ${error.params.allowedValues.join(', ')}` : '';

    return {
      path: error.dataPath.split('.').slice(1),
      code: error.keyword || '',
      message: `${error.message || ''}${allowedParameters}`,
      severity,
    };
  });
};

export const validateAgainstSchema = (
  value: unknown,
  schema: JSONSchema,
  prefix?: string
): Option<IPrismDiagnostic[]> => {
  return tryCatch(() => {
    const validate = ajv.compile(schema);
    const valid = validate(value);
    if (!valid) {
      return convertAjvErrors(validate.errors, DiagnosticSeverity.Error).map(error => {
        const path = prefix ? [prefix, ...error.path] : error.path;
        return Object.assign({}, error, { path });
      });
    }
    return [];
  });
};

export const sequenceValidation = sequenceT(getValidation(getSemigroup<IPrismDiagnostic>()));
export const sequenceOption = sequenceT(option);
