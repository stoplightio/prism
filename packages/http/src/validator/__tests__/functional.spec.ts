import { DiagnosticSeverity, HttpParamStyles, IHttpOperation } from '@stoplight/types';
import { httpInputs, httpOperations, httpOutputs } from '../../__tests__/fixtures';
import { validateInput, validateOutput } from '../index';
import { assertRight, assertLeft } from '@stoplight/prism-core/src/__tests__/utils';

const BAD_INPUT = Object.assign({}, httpInputs[2], {
  body: { name: 'Shopping', completed: 'yes' },
  url: Object.assign({}, httpInputs[2].url, { query: { overwrite: 'true' } }),
  headers: { 'x-todos-publish': 'yesterday' },
});

const GOOD_INPUT = Object.assign({}, httpInputs[2], {
  url: Object.assign({}, httpInputs[0].url, { query: { completed: true } }),
});

const BAD_OUTPUT = Object.assign({}, httpOutputs[1], {
  body: { name: 'Shopping', completed: 'yes' },
  headers: { 'x-todos-publish': 'yesterday', 'content-type': 'application/something' },
});

describe('HttpValidator', () => {
  describe('validateInput()', () => {
    describe('all validations are turned on', () => {
      it('returns validation errors for whole request structure', () => {
        expect(validateInput({ resource: httpOperations[2], element: BAD_INPUT })).toMatchSnapshot();
      });

      describe('when all required params are provided', () => {
        it('returns no validation errors', () => {
          assertRight(validateInput({ resource: httpOperations[0], element: GOOD_INPUT }));
        });
      });
    });

    describe('deprecated keyword validation', () => {
      const resource: IHttpOperation = {
        id: 'abc',
        method: 'get',
        path: '/test',
        responses: [
          {
            code: '200',
          },
        ],
        request: {
          query: [
            {
              style: HttpParamStyles.Form,
              deprecated: true,
              name: 'productId',
            },
          ],
        },
      };

      it('returns warnings', () => {
        assertLeft(
          validateInput({
            resource,
            element: {
              method: 'get',
              url: {
                path: '/test',
                query: {
                  productId: 'abc',
                },
              },
            },
          }),
          error =>
            expect(error).toEqual([
              {
                code: 'deprecated',
                message: 'Query param productId is deprecated',
                path: ['query', 'productId'],
                severity: DiagnosticSeverity.Warning,
              },
            ])
        );
      });

      it('does not return warnings', () => {
        assertRight(
          validateInput({
            resource,
            element: {
              method: 'get',
              url: {
                path: '/test',
                query: {},
              },
            },
          })
        );
      });
    });

    describe('headers validation', () => {
      it('is case insensitive', () => {
        assertRight(
          validateInput({
            resource: {
              method: 'GET',
              path: '/hey',
              responses: [
                {
                  code: '200',
                },
              ],
              id: 'hey',
              request: {
                headers: [
                  {
                    name: 'API_KEY',
                    style: HttpParamStyles.Simple,
                    schema: {
                      type: 'string',
                    },
                    required: true,
                  },
                ],
              },
            },
            element: {
              method: 'get',
              url: {
                path: '/hey',
              },
              headers: {
                api_Key: 'ha',
              },
            },
          })
        );
      });
    });

    describe('query validation', () => {
      it('returns only query validation errors', () => {
        assertLeft(
          validateInput({
            resource: httpOperations[2],
            element: BAD_INPUT,
          }),
          error =>
            expect(error).toContainEqual({
              code: 'pattern',
              message: 'should match pattern "^(yes|no)$"',
              path: ['query', 'overwrite'],
              severity: DiagnosticSeverity.Error,
            })
        );
      });
    });
  });

  describe('validateOutput()', () => {
    describe('all validations are turned on', () => {
      it('returns validation errors for whole request structure', () => {
        expect(validateOutput({ resource: httpOperations[1], element: BAD_OUTPUT })).toMatchSnapshot();
      });
    });
  });
});
