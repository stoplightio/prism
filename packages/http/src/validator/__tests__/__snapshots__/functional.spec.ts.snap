// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HttpValidator validateInput() all validations are turned on returns validation errors for whole request structure 1`] = `
Object {
  "_tag": "Left",
  "left": Array [
    Object {
      "code": "type",
      "message": "should be boolean",
      "path": Array [
        "body",
        "completed",
      ],
      "severity": 0,
    },
    Object {
      "code": "format",
      "message": "should match format \\"date-time\\"",
      "path": Array [
        "header",
      ],
      "severity": 0,
    },
    Object {
      "code": "pattern",
      "message": "should match pattern \\"^(yes|no)$\\"",
      "path": Array [
        "query",
        "overwrite",
      ],
      "severity": 0,
    },
  ],
}
`;

exports[`HttpValidator validateOutput() all validations are turned on returns validation errors for whole request structure 1`] = `
Object {
  "_tag": "Left",
  "left": Array [
    Object {
      "message": "The received media type \\"application/something\\" does not match the ones specified in the current response: application/json, application/xml",
      "severity": 0,
    },
    Object {
      "code": "type",
      "message": "should be boolean",
      "path": Array [
        "body",
        "completed",
      ],
      "severity": 0,
    },
    Object {
      "code": "format",
      "message": "should match format \\"date-time\\"",
      "path": Array [
        "header",
      ],
      "severity": 0,
    },
  ],
}
`;
